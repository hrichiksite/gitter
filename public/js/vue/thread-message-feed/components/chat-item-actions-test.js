const mount = require('../../__test__/vuex-mount');
const { default: ChatItemActions } = require('./chat-item-actions.vue');
const {
  createSerializedMessageFixture,
  createSerializedRoomFixture
} = require('../../__test__/fixture-helpers');

describe('thread-message-feed chat-item-actions', () => {
  const message = createSerializedMessageFixture();

  const extendStore = (userId = message.fromUser.id, isAdmin) => store => {
    store.state.user = { id: userId };
    const room = createSerializedRoomFixture('uri/room', { permissions: { admin: isAdmin } });
    store.state.displayedRoomId = room.id;
    store.state.roomMap = { [room.id]: room };
  };

  it('clicking quote option triggers action', () => {
    const { wrapper, stubbedActions } = mount(ChatItemActions, { message }, extendStore());
    wrapper.find('.js-chat-item-quote-action').trigger('click');
    expect(stubbedActions.threadMessageFeed.quoteMessage).toHaveBeenCalled();
  });

  describe('user is not the author of the message', () => {
    it('matches snapshot', () => {
      const { wrapper } = mount(
        ChatItemActions,
        { message },
        extendStore('someOtherUserId-not-me')
      );
      expect(wrapper.element).toMatchSnapshot();
    });

    it('triggers report action if the report option is clicked', () => {
      const { wrapper, stubbedActions } = mount(
        ChatItemActions,
        { message },
        extendStore('someOtherUserId-not-me')
      );
      // removing the original implementation to prevent an API call
      stubbedActions.threadMessageFeed.reportMessage.mockImplementation(() => {});
      wrapper.find('.js-chat-item-report-action').trigger('click');
      expect(stubbedActions.threadMessageFeed.reportMessage).toHaveBeenCalled();
    });
  });

  describe('user can delete the message', () => {
    it('matches snapshot', () => {
      const { wrapper } = mount(ChatItemActions, { message }, extendStore());
      expect(wrapper.element).toMatchSnapshot();
    });

    it('recognises admin', () => {
      const { wrapper } = mount(
        ChatItemActions,
        { message: message },
        extendStore('someUserId', true)
      );
      expect(wrapper.find('.js-chat-item-delete-action').exists()).toEqual(true);
    });

    it('triggers delete action when delete option is clicked', () => {
      const { wrapper, stubbedActions } = mount(ChatItemActions, { message }, extendStore());
      // removing the original implementation to prevent an API call
      stubbedActions.threadMessageFeed.deleteMessage.mockImplementation(() => {});
      wrapper.find('.js-chat-item-delete-action').trigger('click');
      expect(stubbedActions.threadMessageFeed.deleteMessage).toHaveBeenCalled();
    });
  });

  describe('user can edit the message', () => {
    const testMessage = createSerializedMessageFixture({ sent: new Date().toISOString() });

    it('matches snapshot', () => {
      const { wrapper } = mount(
        ChatItemActions,
        { message: testMessage },
        extendStore(testMessage.fromUser.id)
      );
      expect(wrapper.element).toMatchSnapshot();
    });

    it('triggers edit action when edit option is clicked', () => {
      const { wrapper, stubbedActions } = mount(
        ChatItemActions,
        { message: testMessage },
        extendStore(testMessage.fromUser.id)
      );
      // removing the original implementation to prevent an API call
      stubbedActions.threadMessageFeed.editMessage.mockImplementation(() => {});
      wrapper.find('.js-chat-item-edit-action').trigger('click');
      expect(stubbedActions.threadMessageFeed.editMessage).toHaveBeenCalled();
    });
  });
});
