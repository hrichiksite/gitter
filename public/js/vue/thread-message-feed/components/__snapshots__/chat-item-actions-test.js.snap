// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`thread-message-feed chat-item-actions user can delete the message matches snapshot 1`] = `
<div
  class="chat-item__actions"
>
  <b-popover-stub
    boundary="scrollParent"
    boundarypadding="5"
    customclass="chat-item-actions-popover"
    delay="0"
    fallbackplacement="flip"
    offset="0"
    placement="left"
    target="chat-actions-5d147ea84dad9dfbc522317a"
    title=""
    triggers="click blur"
  >
    <div>
      <button
        class="popover-item__action js-chat-item-quote-action"
        title="Quote this message"
      >
        
        Quote
      
      </button>
    </div>
    <div>
      <div
        class="popover-item__action-disabled"
      >
        
        Edit
      
      </div>
    </div>
    <div>
      <button
        class="popover-item__action js-chat-item-delete-action"
        title="Delete this message"
      >
        
        Delete
      
      </button>
    </div>
    <div>
      <div
        class="popover-item__action-disabled"
      >
        
        Report
      
      </div>
    </div>
  </b-popover-stub>
   
  <button
    class="chat-item-actions-button"
    id="chat-actions-5d147ea84dad9dfbc522317a"
  >
    <i
      class="chat-item__icon icon-ellipsis"
    />
  </button>
</div>
`;

exports[`thread-message-feed chat-item-actions user can edit the message matches snapshot 1`] = `
<div
  class="chat-item__actions"
>
  <b-popover-stub
    boundary="scrollParent"
    boundarypadding="5"
    customclass="chat-item-actions-popover"
    delay="0"
    fallbackplacement="flip"
    offset="0"
    placement="left"
    target="chat-actions-5d147ea84dad9dfbc522317a"
    title=""
    triggers="click blur"
  >
    <div>
      <button
        class="popover-item__action js-chat-item-quote-action"
        title="Quote this message"
      >
        
        Quote
      
      </button>
    </div>
    <div>
      <button
        class="popover-item__action js-chat-item-edit-action"
        title="Edit this message"
      >
        
        Edit
      
      </button>
    </div>
    <div>
      <button
        class="popover-item__action js-chat-item-delete-action"
        title="Delete this message"
      >
        
        Delete
      
      </button>
    </div>
    <div>
      <div
        class="popover-item__action-disabled"
      >
        
        Report
      
      </div>
    </div>
  </b-popover-stub>
   
  <button
    class="chat-item-actions-button"
    id="chat-actions-5d147ea84dad9dfbc522317a"
  >
    <i
      class="chat-item__icon icon-ellipsis"
    />
  </button>
</div>
`;

exports[`thread-message-feed chat-item-actions user is not the author of the message matches snapshot 1`] = `
<div
  class="chat-item__actions"
>
  <b-popover-stub
    boundary="scrollParent"
    boundarypadding="5"
    customclass="chat-item-actions-popover"
    delay="0"
    fallbackplacement="flip"
    offset="0"
    placement="left"
    target="chat-actions-5d147ea84dad9dfbc522317a"
    title=""
    triggers="click blur"
  >
    <div>
      <button
        class="popover-item__action js-chat-item-quote-action"
        title="Quote this message"
      >
        
        Quote
      
      </button>
    </div>
    <div>
      <div
        class="popover-item__action-disabled"
      >
        
        Edit
      
      </div>
    </div>
    <div>
      <div
        class="popover-item__action-disabled"
      >
        
        Delete
      
      </div>
    </div>
    <div>
      <button
        class="popover-item__action js-chat-item-report-action"
        title="Report this message"
      >
        
        Report
      
      </button>
    </div>
  </b-popover-stub>
   
  <button
    class="chat-item-actions-button"
    id="chat-actions-5d147ea84dad9dfbc522317a"
  >
    <i
      class="chat-item__icon icon-ellipsis"
    />
  </button>
</div>
`;
